docs/prefetchfiles-nri-plugin.md

```diff
@@ -0,0 +1,127 @@
+
+## Requirements
+
+- [NRI 2.0](https://github.com/containerd/nri): Which has been integrated into containerd since [v1.7.0](https://github.com/containerd/containerd/tree/v1.7.0-beta.1).
+
+## Workflow
+
+1. Add information such as image reference and URL address containing prefetch files to annotations in pod configuration file.
+2. Run the prefetch plugin to monitor RunPodSandbox events.
+3. The prefetch plugin fetches image reference and URL and forwards them to nydus-snapshotter.
+4. Nydus-snapshotter specifies the prefetch list when starting nydus daemon.
+5. Nydusd completes the mounting of the nydus image.
+
+
+
+
+
+## Modify configuration file 
+
+Modify containerd's toml configuration file to enable NRI.
+```console
+sudo tee -a /etc/containerd/config.toml <<- EOF
+[plugins."io.containerd.nri.v1.nri"]
+    config_file = "/etc/nri/nri.conf"
+    disable = false
+    plugin_path = "/opt/nri/plugins"
+    socket_path = "/var/run/nri.sock"
+EOF
+```
+Containerd will load all NRI plugins in the `plugin_path` directory on startup. If you want to start an NRI plugin manually, please add the following configuration to allow other NRI plugins to connect via `socket_path`.
+
+```console
+sudo tee /etc/nri/nri.conf <<- EOF
+disableConnections: false
+EOF
+```
+
+
+If you want to start the plugin using `pre-connection` mode. You need to write a configuration file and place the plugin's binary file and configuration file in the correct directories. Here is an example of configuration file:
+```console
+sudo tee prefetchfiles-nri-plugin.conf <<- EOF
+# UNIX domain socket address for connection to the nydus-snapshotter API
+socket_address = "/run/containerd-nydus/system.sock"
+EOF
+```
+Then move the files to the correct directories and set permissions. 
+```console
+go build 
+sudo install -D -m 755 ./prefetchfiles-nri-plugin /opt/nri/plugins/03-prefetchfiles-nri-plugin
+sudo install -D -m 755 ./prefetchfiles-nri-plugin.conf /etc/nri/conf.d/03-prefetchfiles-nri-plugin.conf
+```
+
+Restart the containerd service.
+
+```console
+sudo systemctl restart containerd
+```
+When manually starting the prefetch NRI plugin, the socket address can be modified through the command line parameter `socket-addr`. 
+
+
+
+After start the prefetch plugin, it will monitor pod creation events. Note that NRI plugin can only be called from containerd/CRI. So creation a pod using crictl as below.
+```console
+sudo tee pod.yaml <<- EOF
+kind: pod
+metadata:
+  name: wordpress-sandbox
+  namespace: default
+  attempt: 1
+  uid: hdishd83djaidwnduwk28bcsb
+log_directory: /tmp
+annotations:
+ containerd.io/nydus-prefetch: |
+    [
```

Comment:
Format it:

```
[
    {
        "image": "ghcr.io/dragonflyoss/image-service/wordpress:nydus-nightly-v6",
        "prefetch": "http://example.com/api/v1/resource/wordpress",
    }
]
```

Reply:
Does the annotation work with multiple line value?

Reply:
yes, if there are multiple images and prefetchlists in annotations, the code works correctly.